Fix deffile parser to allow comments before header

ikaneshiro opened the pull request:
**Description of the Pull Request (PR):**

This PR modifies tokens generated by the deffile parser to include the `%` before the section name. This allows for reliable identification and separate handling of tokens containing section information vs. header information. This moves away from using `IsValidDefinition()` in `/src/pkg/build/build.go` because there isn't a way to know if a file is a __valid__ definition unless it is parsed in its entirety, which is done by `ParseDefinitionFile()` anyway. This also adds more useful error message if no build source is found in a definition that needs one.

**This fixes or addresses the following GitHub issues:**

- Fixes #2211 


**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](../CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](../CHANGELOG.md) if necessary according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](../CONTRIBUTORS.md)


Attn: @singularity-maintainers

ikaneshiro committed:
test for comments before header and headerless definitions

ikaneshiro committed:
include % in section token to uniquely identify them

ikaneshiro committed:
adding helpful error message for definitions without bootstrap: specified

ikaneshiro committed:
noheader definition should end with a newline

ArangoGutierrez committed:
Merge branch 'master' into deffile_parser_fix

ArangoGutierrez committed:
Merge branch 'master' into deffile_parser_fix

ArangoGutierrez reviewed (dismissed):
LGTM

tested with
```
# a
# very
# verbose comment
Bootstrap: library
# a
# very
# verbose comment
From: alpine:3.7
# a
# very
# verbose comment
%runscript
    echo "$@"
# a
# very
# verbose comment
%post
    echo "Hello from inside the container"
    echo "Install additional software here"
```

GodloveD commented on src/pkg/build/types/parser/deffile_test.go:
```diff
@@ -84,6 +84,7 @@ func TestParseDefinitionFile(t *testing.T) {
 		{"Shub", "../../testdata_good/shub/shub", "../../testdata_good/shub/shub.json"},
 		{"Yum", "../../testdata_good/yum/yum", "../../testdata_good/yum/yum.json"},
 		{"Zypper", "../../testdata_good/zypper/zypper", "../../testdata_good/zypper/zypper.json"},
+		{"NoHeader", "../../testdata_good/noheader/noheader", "../../testdata_good/noheader/noheader.json"},
```
It seems that a def file with no header and another one with comments at the top should be two different tests.

ikaneshiro replied:
Yikes good catch. Ill try to get this panic fixed and more fine grained test cases pushed later tonight

GodloveD reviewed (changes requested):
Looks like this still needs a few more 🤗 and 💋'es.

```
05:01 PM 
ubuntu ~/sandbox
$ cat test2.def 

%post
	apt-get update && apt-get install -y curl

05:01 PM 
ubuntu ~/sandbox
$ sudo singularity build --update test test2.def 
WARNING: Authentication token file not found : Only pulls of public images will succeed
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/sylabs/singularity/src/pkg/build/types/parser.doSections(0xc0004e7030, 0xc000095790, 0x1, 0x40)
	/home/david/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:162 +0x421
github.com/sylabs/singularity/src/pkg/build/types/parser.ParseDefinitionFile(0x1201180, 0xc000506210, 0xc000506210, 0x0, 0x0, 0xc0004e9640, 0x140, 0x140, 0x1085520, 0xc000000300, ...)
	/home/david/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:325 +0x39c
github.com/sylabs/singularity/src/pkg/build.makeDef(0x7ffceb0e9840, 0x9, 0xc0004bec00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/david/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:391 +0x2b5
github.com/sylabs/singularity/src/pkg/build.NewBuild(0x7ffceb0e9840, 0x9, 0x7ffceb0e983b, 0x4, 0x10ff440, 0x3, 0xc0004e0100, 0xc0004e9010, 0x1, 0x1, ...)
	/home/david/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:69 +0x77
github.com/sylabs/singularity/src/cmd/singularity/cli.run(0x19cc3e0, 0xc00019fdd0, 0x2, 0x3)
	/home/david/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/build_linux.go:62 +0x45f
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).execute(0x19cc3e0, 0xc000030070, 0x3, 0x3, 0x19cc3e0, 0xc000030070)
	/home/david/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x19cf5c0, 0x0, 0x68, 0x11f8401)
	/home/david/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).Execute(0x19cf5c0, 0x4, 0x113416c)
	/home/david/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:800 +0x2b
github.com/sylabs/singularity/src/cmd/singularity/cli.ExecuteSingularity()
	/home/david/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/singularity.go:110 +0x91
main.main()
	/home/david/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli.go:16 +0x20

```

Note that the def file above starts with a newline.  Should be fine when used with the `--update` option afaik.

ikaneshiro committed:
add finer grained tests around definitions without headers

ikaneshiro committed:
fix panic caused by definitions with whitespace before the first section

ikaneshiro committed:
Merge branch 'deffile_parser_fix' of github.com:ikaneshiro/singularity into deffile_parser_fix

ikaneshiro committed:
Changed to IsValidDefinition() and removed canGetHeader()

IsValidDefinition() needs to parse the whole file in order to truely determine if it is a valid definition, so the previous method of trying to strip out a header is wrong. This turns IsValidDefiniton() into a boolean wrapper for ParseDefinitionFile().

ikaneshiro committed:
Merge branch 'master' into deffile_parser_fix

GodloveD reviewed (approved):
Sorry it took me a while to get back to this.  lgtm

GodloveD committed:
Merge branch 'master' into deffile_parser_fix

bauerm97 committed:
Merge branch 'master' into deffile_parser_fix
